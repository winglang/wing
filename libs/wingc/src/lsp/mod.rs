mod code_actions;
mod completions;
mod document_symbols;
mod goto_definition;
mod hover;
mod signature;
mod sync;
