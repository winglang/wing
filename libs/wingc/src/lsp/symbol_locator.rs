use crate::{
	ast::*,
	diagnostic::*,
	type_check::{
		resolve_super_method, resolve_user_defined_type_ref, symbol_env::*, ClassLike, Type, TypeRef, Types,
		CLASS_INFLIGHT_INIT_NAME, CLASS_INIT_NAME,
	},
	visit::*,
	visit_context::*,
};

#[derive(Debug)]
/// Result of using the [SymbolLocato]
pub enum SymbolLocatorResult {
	NotFound,
	/// Simple symbol reference
	Symbol(Symbol),
	/// Reference to a property on an object
	ObjectPropertyReference {
		object: ExprId,
		property: Symbol,
		span: WingSpan,
	},
	/// Reference to a property on a type
	TypePropertyReference {
		type_name: UserDefinedType,
		property: Symbol,
		span: WingSpan,
	},
	/// Direct reference to a type
	TypeReference {
		nested_string: String,
		span: WingSpan,
	},
	// Field on a struct
	StructField {
		struct_type: TypeRef,
		field: Symbol,
	},
	/// This is a json key
	LooseField {
		object: TypeRef,
		field_type: TypeRef,
		field: Symbol,
	},
}

pub struct SymbolLocator<'a> {
	ctx: VisitContext,
	types: &'a Types,
	location: WingLocation,
	result_ctx: VisitContext,
	pub result: SymbolLocatorResult,
}

impl<'a> VisitorWithContext for SymbolLocator<'a> {
	fn ctx(&mut self) -> &mut VisitContext {
		&mut self.ctx
	}
}

impl<'a> SymbolLocator<'a> {
	pub fn new(types: &'a Types, location: WingLocation) -> Self {
		Self {
			types,
			ctx: VisitContext::new(),
			result: SymbolLocatorResult::NotFound,
			location,
			result_ctx: VisitContext::new(),
		}
	}

	fn set_result(&mut self, result: SymbolLocatorResult) {
		self.result = result;
		self.result_ctx = self.ctx.clone();
	}

	fn push_scope_env(&mut self, scope: &'a Scope) {
		self.ctx.push_env(self.types.get_scope_env(scope));
	}

	fn is_found(&self) -> bool {
		!matches!(self.result, SymbolLocatorResult::NotFound)
	}

	fn get_env_from_classlike_symbol(&'a self, sym: &'a Symbol) -> Option<SymbolEnvRef> {
		let struct_type = self.ctx.current_env()?.lookup(sym, None)?;
		let struct_type = struct_type.as_type()?;
		match &*struct_type {
			Type::Class(c) => Some(c.get_env().into()),
			Type::Interface(c) => Some(c.get_env().into()),
			Type::Struct(c) => Some(c.get_env().into()),
			_ => None,
		}
	}

	fn lookup_property_on_udt(
		&'a self,
		type_name: &'a UserDefinedType,
		property: &'a Symbol,
	) -> Option<LookupResult<'a>> {
		let current_env = self.result_ctx.current_env()?;
		let type_ = resolve_user_defined_type_ref(type_name, current_env, self.result_ctx.current_stmt_idx()).ok()?;

		match &**type_ {
			// enum variants are not really properties, so just return the enum info
			Type::Enum(..) => {
				let lookup = current_env.lookup_nested_str(&type_name.full_path_str(), Some(self.ctx.current_stmt_idx()));
				Some(lookup)
			}
			_ => self.lookup_property_on_type(type_, property),
		}
	}

	fn lookup_property_on_type(&'a self, type_: &'a TypeRef, property: &'a Symbol) -> Option<LookupResult<'a>> {
		match &**type_ {
			Type::Optional(_)
			| Type::Anything
			| Type::Void
			| Type::Nil
			| Type::Unresolved
			| Type::Inferred(_)
			| Type::Function(_)
			| Type::Enum(_) => None,

			Type::Array(_)
			| Type::MutArray(_)
			| Type::Map(_)
			| Type::MutMap(_)
			| Type::Set(_)
			| Type::MutSet(_)
			| Type::Json(_)
			| Type::MutJson
			| Type::Number
			| Type::String
			| Type::Duration
			| Type::Boolean => {
				if let Some((std_type, ..)) = self.types.get_std_class(&type_) {
					if let Some(t) = std_type.as_type_ref() {
						if let Some(c) = t.as_class() {
							let env = c.get_env();
							let lookup = env.lookup_ext(property, None);
							return Some(lookup);
						}
					}
				}
				None
			}

			Type::Class(c) => {
				let env = c.get_env();
				return Some(env.lookup_ext(property, None));
			}
			Type::Interface(c) => {
				let env = c.get_env();
				return Some(env.lookup_ext(property, None));
			}
			Type::Struct(c) => {
				let env = c.get_env();
				return Some(env.lookup_ext(property, None));
			}
		}
	}

	pub fn lookup_located_symbol(&'a self) -> Option<LookupResult<'a>> {
		match &self.result {
			SymbolLocatorResult::NotFound => None,
			SymbolLocatorResult::Symbol(symbol) => {
				let current_env = self.result_ctx.current_env()?;
				return Some(current_env.lookup_ext(symbol, None));
			}
			SymbolLocatorResult::ObjectPropertyReference { object, property, .. } => {
				let expr_type = self.types.get_expr_id_type_ref(*object).maybe_unwrap_option();

				self.lookup_property_on_type(&expr_type, property)
			}
			SymbolLocatorResult::TypePropertyReference {
				type_name, property, ..
			} => self.lookup_property_on_udt(type_name, property),
			SymbolLocatorResult::TypeReference { nested_string, .. } => {
				let current_env = self.result_ctx.current_env()?;
				let lookup = current_env.lookup_nested_str(&nested_string, None);
				return Some(lookup);
			}
			SymbolLocatorResult::StructField { struct_type, field } => self.lookup_property_on_type(&struct_type, field),
			SymbolLocatorResult::LooseField { .. } => None,
		}
	}

	/// Get the span of the symbol that was located
	pub fn located_span(&self) -> Option<&WingSpan> {
		match &self.result {
			SymbolLocatorResult::NotFound => None,
			SymbolLocatorResult::Symbol(symbol) => Some(&symbol.span),
			SymbolLocatorResult::StructField { field, .. } => Some(&field.span),
			SymbolLocatorResult::LooseField { field, .. } => Some(&field.span),
			SymbolLocatorResult::TypePropertyReference { property, .. }
			| SymbolLocatorResult::ObjectPropertyReference { property, .. } => Some(&property.span),
			SymbolLocatorResult::TypeReference { span, .. } => Some(&span),
		}
	}
}

impl<'a> Visit<'a> for SymbolLocator<'a> {
	fn visit_scope(&mut self, node: &'a Scope) {
		if self.is_found() {
			return;
		}

		self.ctx.push_env(self.types.get_scope_env(node));
		visit_scope(self, node);
		self.ctx.pop_env();
	}

	fn visit_stmt(&mut self, node: &'a crate::ast::Stmt) {
		if self.is_found() {
			return;
		}

		self.ctx.push_stmt(node.idx);

		// Handle situations where symbols are actually defined in inner scopes
		match &node.kind {
			StmtKind::Struct(st) => {
				let Some(struct_env) = self.get_env_from_classlike_symbol(&st.name) else {
					return;
				};

				self.ctx.push_env(struct_env);

				for field in &st.fields {
					self.visit_symbol(&field.name);
				}

				self.ctx.pop_env();
			}
			StmtKind::Interface(interface) => {
				let Some(interface_env) = self.get_env_from_classlike_symbol(&interface.name) else {
					return;
				};

				self.ctx.push_env(interface_env);

				for method in &interface.methods {
					self.visit_symbol(&method.0);
				}

				self.ctx.pop_env();
			}
			StmtKind::ForLoop {
				iterator, statements, ..
			} => {
				self.push_scope_env(&statements);
				self.visit_symbol(iterator);
				self.ctx.pop_env();
			}
			StmtKind::TryCatch { catch_block, .. } => {
				if let Some(catch_block) = catch_block {
					if let Some(exception_var) = &catch_block.exception_var {
						self.push_scope_env(&catch_block.statements);
						self.visit_symbol(exception_var);
						self.ctx.pop_env();
					}
				}
			}
			StmtKind::IfLet(IfLet {
				var_name,
				statements,
				elif_statements,
				..
			}) => {
				self.push_scope_env(&statements);
				self.visit_symbol(var_name);
				self.ctx.pop_env();

				for elif in elif_statements {
					if let Elifs::ElifLetBlock(elif_let_block) = elif {
						self.push_scope_env(&elif_let_block.statements);
						self.visit_symbol(&elif_let_block.var_name);
						self.ctx.pop_env();
					}
				}
			}
			_ => {}
		}

		visit_stmt(self, node);

		self.ctx.pop_stmt();
	}

	fn visit_expr(&mut self, node: &'a Expr) {
		if self.is_found() {
			return;
		}

		self.ctx.push_expr(node.id);

		match &node.kind {
			ExprKind::New(new_expr) => {
				let found_named_arg = new_expr
					.arg_list
					.named_args
					.iter()
					.find(|a| a.0.span.contains_location(&self.location));

				if let Some((arg_name, ..)) = found_named_arg {
					// we need to get the struct type from the class constructor
					let class_type = self.types.get_expr_type(node);
					let Some(class_phase) = self.types.get_expr_phase(node) else {
						return;
					};
					let Some(class_type) = class_type.as_class() else {
						return;
					};
					let init_info = match class_phase {
						Phase::Inflight => class_type.get_method(&Symbol::global(CLASS_INFLIGHT_INIT_NAME)),
						Phase::Preflight => class_type.get_method(&Symbol::global(CLASS_INIT_NAME)),
						Phase::Independent => panic!("Cannot get hover info for independent class"),
					};
					if let Some(var_info) = init_info {
						if let Some(func) = var_info.type_.maybe_unwrap_option().as_function_sig() {
							if let Some(arg) = func.parameters.last() {
								let struct_type = arg.typeref.maybe_unwrap_option();
								self.set_result(SymbolLocatorResult::StructField {
									struct_type: struct_type.to_owned(),
									field: arg_name.clone(),
								});
							}
						}
					}
				}
			}
			ExprKind::Call { arg_list, callee } => {
				let x = arg_list
					.named_args
					.iter()
					.find(|a| a.0.span.contains_location(&self.location));
				if let Some((arg_name, ..)) = x {
					let Some(env) = self.ctx.current_env() else {
						return;
					};
					// we need to get the struct type from the callee
					let callee_type = match callee {
						CalleeKind::Expr(expr) => self.types.get_expr_type(expr),
						CalleeKind::SuperCall(method) => resolve_super_method(method, &env, &self.types)
							.ok()
							.map_or(self.types.error(), |t| t.0),
					}
					.maybe_unwrap_option()
					.to_owned();

					if let Some(func) = callee_type.as_function_sig() {
						if let Some(arg) = func.parameters.last() {
							let struct_type = arg.typeref.maybe_unwrap_option();
							self.set_result(SymbolLocatorResult::StructField {
								struct_type: struct_type.to_owned(),
								field: arg_name.clone(),
							});
						}
					}
				}
			}
			ExprKind::JsonMapLiteral { fields } | ExprKind::StructLiteral { fields, .. } => {
				if let Some(f) = fields.iter().find(|f| f.0.span.contains_location(&self.location)) {
					let field_name = f.0;
					let type_ = self.types.maybe_unwrap_inference(self.types.get_expr_type(node));
					let type_ = *if let Some(type_) = self.types.get_type_from_json_cast(node.id) {
						*type_
					} else {
						type_
					}
					.maybe_unwrap_option();

					if type_.is_struct() {
						self.set_result(SymbolLocatorResult::StructField {
							struct_type: type_.to_owned(),
							field: field_name.clone(),
						});
					} else {
						// just use the type info
						let inner_type = self.types.maybe_unwrap_inference(self.types.get_expr_type(f.1));
						self.set_result(SymbolLocatorResult::LooseField {
							object: type_.to_owned(),
							field_type: inner_type.to_owned(),
							field: field_name.clone(),
						});
					}
				}
			}
			_ => {}
		}

		crate::visit::visit_expr(self, node);

		self.ctx.pop_expr();
	}

	fn visit_function_definition(&mut self, node: &'a FunctionDefinition) {
		if self.is_found() {
			return;
		}

		if let FunctionBody::Statements(scope) = &node.body {
			self.push_scope_env(scope);
			for param in &node.signature.parameters {
				self.visit_function_parameter(&param);
			}
			self.ctx.pop_env();
		}

		visit_function_definition(self, node);
	}

	fn visit_class(&mut self, node: &'a Class) {
		if self.is_found() {
			return;
		}

		let Some(class_env) = self.get_env_from_classlike_symbol(&node.name) else {
			return;
		};

		self.ctx.push_env(class_env);

		for field in &node.fields {
			self.visit_symbol(&field.name);
		}
		for method in &node.methods {
			self.visit_symbol(&method.0);
		}

		self.ctx.pop_env();

		visit_class(self, node);
	}

	fn visit_reference(&mut self, node: &'a Reference) {
		if self.is_found() {
			return;
		}

		match node {
			Reference::Identifier(sym) => self.visit_symbol(sym),
			Reference::InstanceMember { object, property, .. } => {
				self.visit_expr(object);

				if property.span.contains_location(&self.location) {
					self.set_result(SymbolLocatorResult::ObjectPropertyReference {
						span: object.span.merge(&property.span),
						object: object.id,
						property: property.clone(),
					});
				}
			}
			Reference::TypeMember { type_name, property } => {
				if property.span.contains_location(&self.location) {
					self.set_result(SymbolLocatorResult::TypePropertyReference {
						type_name: type_name.clone(),
						property: property.clone(),
						span: type_name.span.merge(&property.span),
					});
				} else {
					self.visit_user_defined_type(type_name);
				}
			}
			Reference::ElementAccess { object, .. } => {
				self.visit_expr(object);
			}
		}

		visit_reference(self, node);
	}

	fn visit_user_defined_type(&mut self, node: &'a UserDefinedType) {
		if self.is_found() {
			return;
		}

		if node.span.contains_location(&self.location) {
			if node.root.span.contains_location(&self.location) {
				self.visit_symbol(&node.root);
			} else {
				let mut full_string = node.root.name.clone();
				let mut full_span = node.root.span.clone();
				for segment in &node.fields {
					full_string = format!("{}.{}", full_string, segment.name);
					full_span = full_span.merge(&segment.span);

					if segment.span.contains_location(&self.location) {
						self.set_result(SymbolLocatorResult::TypeReference {
							nested_string: full_string,
							span: full_span,
						});
						break;
					}
				}
			}
		}
	}

	fn visit_symbol(&mut self, node: &'a crate::ast::Symbol) {
		if self.is_found() {
			return;
		}

		if node.span.contains_location(&self.location) {
			self.set_result(SymbolLocatorResult::Symbol(node.clone()));
		}
	}
}
