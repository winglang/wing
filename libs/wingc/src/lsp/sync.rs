use camino::{Utf8Path, Utf8PathBuf};
use indexmap::IndexMap;
use lsp_types::{DidChangeTextDocumentParams, DidOpenTextDocumentParams};
use wingii::type_system::TypeSystem;

use std::cell::RefCell;
use std::path::{Path, PathBuf};
use tree_sitter::Tree;

use crate::closure_transform::ClosureTransformer;
use crate::diagnostic::{found_errors, reset_diagnostics};
use crate::file_graph::FileGraph;
use crate::files::Files;
use crate::fold::Fold;
use crate::jsify::JSifier;
use crate::lifting::LiftVisitor;
use crate::parser::{normalize_path, parse_wing_project};
use crate::type_check;
use crate::type_check::jsii_importer::JsiiImportSpec;
use crate::type_check::type_reference_transform::TypeReferenceTransformer;
use crate::type_check_assert::TypeCheckAssert;
use crate::valid_json_visitor::ValidJsonVisitor;
use crate::visit::Visit;
use crate::wasm_util::extern_json_fn;
use crate::{ast::Scope, type_check::Types};

/// The output of compiling a Wing project with one or more files
pub struct ProjectData {
	/// Text data contained in the files (utf8)
	pub files: Files,
	/// A graph that tracks the dependencies between files
	pub file_graph: FileGraph,
	/// tree-sitter trees
	pub trees: IndexMap<Utf8PathBuf, Tree>,
	/// AST for each file
	pub asts: IndexMap<Utf8PathBuf, Scope>,
	/// The JSII imports for the file. This is saved so we can load JSII types (for autocompletion for example)
	/// which don't exist explicitly in the source.
	pub jsii_imports: Vec<JsiiImportSpec>,
}

impl ProjectData {
	fn new() -> Self {
		ProjectData {
			files: Files::new(),
			file_graph: FileGraph::default(),
			trees: IndexMap::new(),
			asts: IndexMap::new(),
			jsii_imports: Vec::new(),
		}
	}
}

thread_local! {
	/// When consumed as a WASM library, wingc is not in control of the process/memory in which it is running.
	/// This means that it cannot reliably manage stateful data like this between function calls.
	/// Here we will assume the process is single threaded, and use thread_local to store this data.
	pub static WING_TYPES: RefCell<Types> = RefCell::new(Types::new());
	pub static PROJECT_DATA: RefCell<ProjectData> = RefCell::new(ProjectData::new());
	pub static JSII_TYPES: RefCell<TypeSystem> = RefCell::new(TypeSystem::new());
}

#[no_mangle]
pub unsafe extern "C" fn wingc_on_did_open_text_document(ptr: u32, len: u32) {
	extern_json_fn(ptr, len, on_document_did_open);
}

pub fn on_document_did_open(params: DidOpenTextDocumentParams) {
	WING_TYPES.with(|wing_types| {
		JSII_TYPES.with(|jsii_types| {
			PROJECT_DATA.with(|project_data| {
				let uri = params.text_document.uri;
				let uri_path = uri.to_file_path().unwrap();
				let source_text = params.text_document.text;

				partial_compile(
					&uri_path,
					source_text,
					&mut wing_types.borrow_mut(),
					&mut jsii_types.borrow_mut(),
					&mut project_data.borrow_mut(),
				);
			});
		});
	});
}

#[no_mangle]
pub unsafe extern "C" fn wingc_on_did_change_text_document(ptr: u32, len: u32) {
	extern_json_fn(ptr, len, on_document_did_change);
}

pub fn on_document_did_change(params: DidChangeTextDocumentParams) {
	let DidChangeTextDocumentParams {
		text_document,
		content_changes,
	} = params;

	WING_TYPES.with(|wing_types| {
		JSII_TYPES.with(|jsii_types| {
			PROJECT_DATA.with(|project_data| {
				let uri = text_document.uri;
				let uri_path = uri.to_file_path().unwrap();
				let source_text = content_changes.into_iter().next().unwrap().text;

				partial_compile(
					&uri_path,
					source_text,
					&mut wing_types.borrow_mut(),
					&mut jsii_types.borrow_mut(),
					&mut project_data.borrow_mut(),
				);
			});
		})
	});
}

/// Runs several phases of the wing compiler on a file, including: parsing, type checking, and lifting
/// `ProjectData` is passed with results from previous compilations, and is updated with the results of this compilation.
fn partial_compile(
	source_path: &Path,
	source_text: String,
	mut types: &mut Types, // TODO: does this need to be shared between recompiles?
	jsii_types: &mut TypeSystem,
	project_data: &mut ProjectData,
) {
	// Reset diagnostics before new compilation (`partial_compile` can be called multiple times)
	reset_diagnostics();

	let source_path = Utf8Path::from_path(source_path).expect("invalid unicode path");
	let source_path = normalize_path(source_path, None);

	let topo_sorted_files = parse_wing_project(
		&source_path,
		Some(source_text),
		&mut project_data.files,
		&mut project_data.file_graph,
		&mut project_data.trees,
		&mut project_data.asts,
	);

	// -- DESUGARING PHASE --

	// Transform all inflight closures defined in preflight into single-method resources
	for file in &topo_sorted_files {
		let mut inflight_transformer = ClosureTransformer::new();
		let scope = project_data.asts.remove(file).unwrap();
		let new_scope = inflight_transformer.fold_scope(scope);
		project_data.asts.insert(file.clone(), new_scope);
	}

	// Reset all type information
	types.reset_expr_types();
	types.reset_scope_envs();

	// -- TYPECHECKING PHASE --
	let mut jsii_imports = vec![];

	// Type check all files in topological order (start with files that don't require any other
	// Wing files, then move on to files that depend on those, etc.)
	for file in &topo_sorted_files {
		let mut scope = project_data.asts.remove(file).expect("matching AST not found");
		type_check(
			&mut scope,
			&mut types,
			&file,
			&project_data.file_graph,
			jsii_types,
			&mut jsii_imports,
		);

		// Make sure all type reference are no longer considered references
		let mut tr_transformer = TypeReferenceTransformer { types: &mut types };
		let scope = tr_transformer.fold_scope(scope);

		// Validate the type checker didn't miss anything - see `TypeCheckAssert` for details
		let mut tc_assert = TypeCheckAssert::new(&types, found_errors());
		tc_assert.check(&scope);

		// Validate all Json literals to make sure their values are legal
		let mut json_checker = ValidJsonVisitor::new(&types);
		json_checker.check(&scope);

		project_data.asts.insert(file.clone(), scope);
	}

	// -- LIFTING PHASE --

	let jsifier = JSifier::new(
		&mut types,
		&project_data.files,
		&project_data.file_graph,
		&source_path,
		// out_dir will not be used
		&source_path,
	);
	for file in &topo_sorted_files {
		let mut lift = LiftVisitor::new(&jsifier);
		let scope = project_data.asts.remove(file).expect("matching AST not found");
		lift.visit_scope(&scope);
		project_data.asts.insert(file.clone(), scope);
	}

	// no need to JSify in the LSP
}

pub fn check_utf8(path: PathBuf) -> Utf8PathBuf {
	path.try_into().expect("invalid unicode path")
}

#[cfg(test)]
pub mod test_utils {
	use std::{fs, str::FromStr};
	use uuid::Uuid;

	use lsp_types::*;

	use crate::diagnostic::assert_no_panics;

	use super::on_document_did_open;

	static LANGUAGE_ID: &str = "wing";

	/// Loads a file into the language server
	///
	/// Returns a `TextDocumentPositionParams` with the generated URI and a cursor position of an indicated position in the file (using `//^`)
	///
	/// Example that gives you the position of the "o" in "Resource.":
	/// ```
	/// load_file_with_contents(
	/// r#"
	/// class Resource {
	/// 	static hello() {}
	/// }
	///
	/// Resource.
	///  //^"#
	/// )
	/// ```
	///
	pub fn load_file_with_contents(content: &str) -> TextDocumentPositionParams {
		let temp_dir = tempfile::tempdir().expect("Failed to create temporary directory");
		let filename = format!("{}.main.w", Uuid::new_v4());
		let file_path = temp_dir.path().join(&filename);
		fs::write(&file_path, content).expect("Failed to write to temporary file");
		let file_uri_string = format!("file:///{}", file_path.to_str().unwrap());
		let uri = Url::from_str(&file_uri_string).unwrap();
		on_document_did_open(DidOpenTextDocumentParams {
			text_document: lsp_types::TextDocumentItem {
				uri: uri.clone(),
				language_id: LANGUAGE_ID.to_string(),
				version: 0,
				text: content.to_string(),
			},
		});

		assert_no_panics();

		// find the character cursor position by looking for the character above the ^
		let mut char_pos = 0_i32;
		let mut line_pos = 0_i32;
		// Note: `.chars()` may not 1-to-1 map with characters as expected if using things like emojis
		// This generally acceptable for testing, but should not be assumed to be the case in the real language server
		// TODO: Add support for grapheme clusters
		for char in content.chars() {
			if char == '^' {
				break;
			} else if char == '\n' {
				char_pos = 0;
				line_pos += 1;
			} else {
				char_pos += 1;
			}
		}

		let cursor_position = Position {
			line: (line_pos - 1).max(0) as u32,
			character: char_pos as u32,
		};

		return TextDocumentPositionParams {
			text_document: TextDocumentIdentifier { uri },
			position: cursor_position,
		};
	}

	/// Finds all ranges in the document starting with `//-`
	pub fn get_ranges(content: &str) -> Vec<Range> {
		let lines = content.lines();

		let mut ranges = vec![];
		for line in lines.enumerate() {
			if line.1.contains("//-") {
				let start_col = line.1.match_indices("//-").next().unwrap().0 + 2;
				let end_col = line.1.len();
				let line_num = line.0 - 1;

				ranges.push(Range {
					start: Position {
						line: line_num as u32,
						character: start_col as u32,
					},
					end: Position {
						line: line_num as u32,
						character: end_col as u32,
					},
				});
			}
		}

		ranges
	}
}
