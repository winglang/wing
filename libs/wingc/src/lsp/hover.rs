use crate::docs::Documented;
use crate::lsp::sync::PROJECT_DATA;
use crate::type_check::symbol_env::LookupResult;
use crate::visit::Visit;
use crate::wasm_util::extern_json_fn;
use lsp_types::{Hover, HoverContents, MarkupContent, MarkupKind};

use super::symbol_locator::{SymbolLocator, SymbolLocatorResult};
use super::sync::{check_utf8, WING_TYPES};

#[no_mangle]
pub unsafe extern "C" fn wingc_on_hover(ptr: u32, len: u32) -> u64 {
	extern_json_fn(ptr, len, on_hover)
}
pub fn on_hover(params: lsp_types::HoverParams) -> Option<Hover> {
	WING_TYPES.with(|types| {
		let types = types.borrow_mut();
		PROJECT_DATA.with(|project_data| {
			let project_data = project_data.borrow();
			let uri = params.text_document_position_params.text_document.uri.clone();
			let file = check_utf8(uri.to_file_path().expect("LSP only works on real filesystems"));
			let root_scope = &project_data.asts.get(&file).unwrap();

			let mut symbol_finder = SymbolLocator::new(&types, params.text_document_position_params.position.into());
			symbol_finder.visit_scope(root_scope);

			if let Some(lookup) = symbol_finder.lookup_located_symbol() {
				if let LookupResult::Found(symbol_info, ..) = &lookup {
					let docs = symbol_info.render_docs();
					let span = symbol_finder.located_span()?;
					return Some(Hover {
						contents: HoverContents::Markup(MarkupContent {
							kind: MarkupKind::Markdown,
							value: docs,
						}),
						range: Some(span.into()),
					});
				}
			} else {
				if let SymbolLocatorResult::LooseField { field_type, field, .. } = &symbol_finder.result {
					let span = symbol_finder.located_span().unwrap();
					let value = format!("```wing\n{}: {field_type}\n```", field.name);

					return Some(Hover {
						contents: HoverContents::Markup(MarkupContent {
							kind: MarkupKind::Markdown,
							value,
						}),
						range: Some(span.into()),
					});
				}
			}

			None
		})
	})
}

#[cfg(test)]
mod tests {
	use crate::lsp::hover::*;
	use crate::lsp::sync::test_utils::*;
	use lsp_types::*;

	/// Creates a snapshot test for a given wing program's hover at a given position
	/// In the wing program, place a comment "//^" into the text where the "^" is pointing to the desired character position
	///
	/// First parameter will be the name of the tests, as well as the identifier to use for the list of completion in the asserts (see last parameter)
	/// Second parameter is the wing code block as a string literal
	/// After the first two parameters, any additional are optional statements that should be used for asserting on the given hover data.
	///
	/// Result is a [Hover] object
	macro_rules! test_hover_list {
		($name:ident, $code:literal) => {
			test_hover_list!($name, $code,);
		};
		($name:ident, $code:literal, $($assertion:stmt)*) => {
			#[test]
			fn $name() {
				// NOTE: this is needed for debugging to work regardless of where you run the test
				std::env::set_current_dir(env!("CARGO_MANIFEST_DIR")).unwrap();

				let text_document_position_params = load_file_with_contents($code);
				let hover = on_hover(HoverParams {
					text_document_position_params,
					work_done_progress_params: Default::default(),
				});

				if let Some($name) = hover {
					insta::with_settings!(
						{
							prepend_module_to_snapshot => false,
							omit_expression => true,
							snapshot_path => "./snapshots/hovers",
						}, {
							insta::assert_yaml_snapshot!($name);
						}
					);
					$($assertion)*
				} else {
					panic!("Expected hover data");
				}
			}
		};
	}

	test_hover_list!(
		just_variable,
		r#"
		let myString = "hello";
        //^
		"#,
	);

	test_hover_list!(
		new_expression_nested,
		r#"
bring cloud;
new cloud. 
		//^
"#,
	);

	test_hover_list!(
		class_symbol,
		r#"
		bring cloud;

		let bucket = new cloud.Bucket();
        //^
"#,
	);

	test_hover_list!(
		class_symbol_in_closure,
		r#"
inflight class MyClass { }

inflight () => {
  let myClass = new MyClass();
    //^

}
"#,
	);

	test_hover_list!(
		class_property,
		r#"
bring cloud;

let bucket = new cloud.Bucket();
bucket.addObject
      //^
"#,
	);

	test_hover_list!(
		static_stdtype_method,
		r#"
Json.stringify(123);
      //^
"#,
	);

	test_hover_list!(
		inside_inflight_test,
		r#"
test "hello" {
  let a = 1;
    //^
}
"#,
	);

	test_hover_list!(
		inside_class_method,
		r#"
class Foo {
  bar() {
    let hello = 1;
        //^
  }
}
"#,
	);

	test_hover_list!(
		inside_class_field,
		r#"
bring cloud;
class Foo {
  my_bucket: cloud.Bucket;
    //^
}
"#,
	);

	test_hover_list!(
		new_statement,
		r#"
bring cloud;
new cloud.Bucket();
          //^
"#
	);

	test_hover_list!(
		user_defined_types,
		r#"
class Foo { }
     //^
"#
	);

	test_hover_list!(
		user_defined_type_annotation,
		r#"
class Foo { }
let a: Foo = new Foo();
      //^
"#
	);

	test_hover_list!(
		user_defined_type_reference_property,
		r#"
class Foo { 
	static static_method() { }
}
Foo.static_method();
   //^
"#
	);

	test_hover_list!(
		user_defined_type_reference_type,
		r#"
class Foo { 
	static static_method() { }
}
Foo.static_method();
//^
"#
	);

	test_hover_list!(
		static_method,
		r#"
class Foo {
  static my(a: str, b: bool): str { return "str"; }
}

Foo.my();
  //^
"#
	);

	test_hover_list!(
		builtin_in_preflight,
		r#"
assert(true);
//^
"#
	);

	test_hover_list!(
		builtin_in_inflight,
		r#"
class Foo {
  inflight bar() {
    assert(true);
    //^
  }
}
"#
	);

	test_hover_list!(
		test_bring_sdk,
		r#"
bring cloud;
      //^
"#
	);

	test_hover_list!(
		test_var,
		r#"
let var xoo = "hello";
log(xoo);
    //^
"#
	);

	test_hover_list!(
		test_var_inside_preflight_closure,
		r#"
() => {
  let var goooo = "gar";
          //^
}"#
	);

	test_hover_list!(
		test_var_inside_inflight_closure,
		r#"
inflight () => {
  let var goooo = "gar";
          //^
}"#
	);

	test_hover_list!(
		test_builtin_instance_method,
		r#"
"hello".startsWith("h");
           //^
"#
	);

	test_hover_list!(
		multipart_reference_hover_middle,
		r#"
let j = Json {};
j.get("hello").get("world");
 //^
"#
	);

	test_hover_list!(
		map_element,
		r#"
let k = "hi";
{ k => "" }
//^
"#
	);

	test_hover_list!(
		json_element,
		r#"
{ hi: "cool" }
 //^
"#
	);

	test_hover_list!(
		json_element_nested_top,
		r#"
{ hi: { inner: [1, 2, 3] } }
 //^
"#
	);

	test_hover_list!(
		json_element_nested_inner,
		r#"
{ hi: { inner: [1, 2, 3] } }
        //^
"#
	);

	test_hover_list!(
		inflight_init,
		r#"
struct Data {
	field: str;
}

class T {
	new() {
		Data { field: "" };
	}

	inflight new() {
		Data { field: "" };
		//^
	}
}
"#
	);

	test_hover_list!(
		class_init_this_field,
		r#"
class T {
  stuff: num;

  new() {
    this.stuff = 1;
         //^
  }
}
"#
	);

	test_hover_list!(
		struct_decl_field,
		r#"
struct S {
  field: str;
  //^
}
"#
	);

	test_hover_list!(
		static_method_root,
		r#"
Json.stringify({});
//^
"#
	);

	test_hover_list!(
		variadic_args,
		r#"
  class Arr {
    pub addMany(...items: Array<str>) {
  
    }
  }
  
  let arr = new Arr();
  arr.addMany("a","b","c");
      //^
  "#,
	);
}
