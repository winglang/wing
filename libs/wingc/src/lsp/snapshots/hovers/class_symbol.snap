---
source: libs/wingc/src/lsp/hover.rs
---
contents:
  kind: markdown
  value: "```wing\npreflight bucket: Bucket\nclass Bucket\n```\n---\nA cloud object store."
range:
  start:
    line: 3
    character: 6
  end:
    line: 3
    character: 12

